lib/Dialect/FIRRTL/Transforms/CheckCombLoops.cpp

```diff
@@ -25,533 +26,472 @@
+                children.push_back(childVal);
+            }
+          }
+          for (auto childVal : children) {
+            // This childVal can be ignored, if
+            // It is a Register or a subfield of a register.
+            if (!visited.contains(childVal))
+              dfsStack.push_back(childVal);
+            // If the childVal is a sub, then check if it aliases with any of
+            // the predecessors (the visiting set).
+            if (visiting.contains(childVal)) {
+              // Comb Cycle Detected !!
+              reportLoopFound(childVal, visiting);
+              return failure();
+            }
+          }
+          // child nodes added, continue the DFS
+          if (dfsSize != dfsStack.size())
+            continue;
         }
-      } else
-        worklist.push_back(Node(res, fieldId));
-    };
-    // Traverse the IR, to get all the lowered fields of the aggregate type.
-    // This representation is equivalent to the LowerTypes transformation.
-    while (!worklist.empty()) {
-      Node node = worklist.pop_back_val();
-      for (auto &use : node.val.getUses()) {
-        Operation *op = use.getOwner();
-        TypeSwitch<Operation *>(op)
-            .Case<SubfieldOp>([&](SubfieldOp sub) {
-              auto bundleType = sub.getInput().getType();
-              auto fId = bundleType.getFieldID(sub.getFieldIndex());
-              auto res = sub.getResult();
-              handleSub(res, node.fieldId + fId);
-            })
-            .Case<SubindexOp>([&](SubindexOp sub) {
-              auto vecType = sub.getInput().getType().cast<FVectorType>();
-              auto res = sub.getResult();
-              auto fId = vecType.getFieldID(sub.getIndex());
-              handleSub(res, node.fieldId + fId);
-            })
-            .Case<SubaccessOp>([&](SubaccessOp sub) {
-              auto vecType = sub.getInput().getType().cast<FVectorType>();
-              auto res = sub.getResult();
-              // The result of a subaccess can refer to multiple storage
-              // locations corresponding to all the possible indices.
-              for (size_t index = 0; index < vecType.getNumElements(); ++index)
-                handleSub(
-                    res,
-                    node.fieldId + 1 +
-                        index * (vecType.getElementType().getMaxFieldID() + 1));
-            })
-            .Default([&](auto op) {});
+        // Node is an SCC root node, pop the visiting stack to remove the nodes
+        // that are no longer active predecessors, that is their sub-tree is
+        // already explored.
+        // All the Values reachable from `dfsVal` have been explored, remove
+        // it and its children from the visiting stack.
+        visiting.popUntillVal(dfsVal);
+
+        auto popped = dfsStack.pop_back_val();
+        LLVM_DEBUG(llvm::dbgs()
+                   << "\n dfs popped :" << getFieldName(Node(popped, 0)).first);
+        dump();
```

Comment:
nit: can we change to `LLVM_DEBUG (dump());` and remove LLVM_DEBUG in `dump`? It gave me an impression that debug message is remaining.